pallets/pallet-asset-swap/src/traits.rs

```diff
@@ -0,0 +1,75 @@
+// KILT Blockchain – https://botlabs.org
+// Copyright (C) 2019-2024 BOTLabs GmbH
+
+// The KILT Blockchain is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The KILT Blockchain is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// If you feel like getting in touch with us, you can do so at info@botlabs.org
+
+use xcm::VersionedMultiLocation;
+
+use crate::{Config, LocalCurrencyBalanceOf};
+
+pub trait SwapHooks<T>
+where
+	T: Config,
+{
+	type Error: Into<u8>;
+
+	fn pre_local_to_remote_swap(
+		from: &T::AccountId,
+		to: &VersionedMultiLocation,
+		amount: LocalCurrencyBalanceOf<T>,
+	) -> Result<(), Self::Error>;
+
+	fn post_local_to_remote_swap(
+		from: &T::AccountId,
+		to: &VersionedMultiLocation,
+		amount: LocalCurrencyBalanceOf<T>,
+	) -> Result<(), Self::Error>;
+
+	fn pre_remote_to_local_swap(to: &T::AccountId, amount: u128) -> Result<(), Self::Error>;
+
+	fn post_remote_to_local_swap(to: &T::AccountId, amount: u128) -> Result<(), Self::Error>;
```

Comment:
I also wonder, do we really need that? Please correct me if I am wrong, but it seems to me those functions are not directly used in the pallet itself, only in some XCM related structs. If someone wants to do some pre- and post-remote to local checks, they can implement it by themselves or not?

Reply:
It doesn't hurt if they would remain. It just feels a bit strange to me.

Reply:
I mean, sure, you can find the right XCM component to do what you want to do. Or simply call into the post hook after everything else happens. I think, as you also say, that it's nice to have pre and post for both local -> remote and remote -> local. There might be cases where you might want to do something else with this information, and unnecessarily restricting it would be a wasted opportunity, I think.